### PR TITLE
Correctly infer branch names for SSO users

### LINKED ISSUE #17681: Opening issue contexts can fail on invalid characters
### Bug description

When an issue context is passed to create a workspace, our server first creates a branch name out of the issue title and the name of the user.

The issue's birthplace is here:
https://github.com/gitpod-io/gitpod/blob/94342d06ffee9c4d001527250470addcdd4fd00d/components/server/src/workspace/context-parser.ts#L112

The exact issue is that the branch name is correctly split and re-added to the branch name as segments joined with a proper deliminator (in our case `-`), but this does not happen to the name of the user. Therefore for me, "Filip Troníček", the branch will have a whitespace character in it, which Git doesn't like. 

### Steps to reproduce

Open a workspace from any issue and observe the error. 

### Workspace affected

_No response_

### Expected behavior

The name is also concatenated by hyphens, just like the issue title. 

### Example repository

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by filiptronicek**
## Description

Makes sure branches from issue contexts are always spec-compliant and therefore valid for git. 

Fixes WEB-375

## How to test

Testing should be done by logging in through SSO and starting a workspace from an issue context. I unfortunately couldn't find a way to do the SSO login, but for proper testing it has to be done like that.

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - ft-fix-issd380c132cf</li>
	<li><b>🔗 URL</b> - <a href="https://ft-fix-issd380c132cf.preview.gitpod-dev.com/workspaces" target="_blank">ft-fix-issd380c132cf.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
</ul>

## Build Options:

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [x] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

/hold

### CODE REVIEW

File: components/server/src/workspace/context-parser.ts [hunk 1]
```diff
@@ -102,6 +102,15 @@ export interface IPrefixContextParser {
...
 
+// See https://www.kernel.org/pub/software/scm/git/docs/git-check-ref-format.html
+// the magic sequence @{, consecutive dots, leading and trailing dot, ref ending in .lock
+// Adapted https://github.com/desktop/desktop/blob/1e3df9608a834dabdabe793b1b538e334f33c8a1/app/src/lib/sanitize-ref-name.ts
+const invalidCharacterRegex = /[\x00-\x20\x7F~^:?*\[\\|""<>]+|@{|\.\.+|^\.|\.$|\.lock$|\/$/g;
+/** Sanitize a proposed reference name by replacing illegal characters. */
+
+function sanitizedRefName(name: string): string {
+    return name.replace(invalidCharacterRegex, "-").replace(/^[-\+]*/g, "");
+}
```
Comment (filiptronicek) [inline components/server/src/workspace/context-parser.ts:113]: Can anyone help me to get this function to a more proper place? I'm not too familiar with the server codebase and would appreciate a tip

**Review by easyCZ — COMMENTED** [review]
Would you be able to add a unit test for at least the case which failed for you and this is a fix for it?

File: components/server/src/workspace/context-parser.ts (hunk 1 above)
Comment (easyCZ) [reply components/server/src/workspace/context-parser.ts:113]: I think here is fine actually. We can move it, should we need to reuse

**Review by easyCZ — APPROVED** [review]

**Pull request merged**